### DISCUSSION TITLE
Longer names are not merged, only last part shown

### DISCUSSION THREAD

**Discussion by kristof-mattei** [zwave-js/zwave-js#7415] [Request Support / Investigate Issue] [answered]
### Checklist

- [X] I have read and followed the above instructions
- [X] I have checked the [troubleshooting section](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index) and my problem is **not** described there.

- [X] I have read the [changelog](https://github.com/zwave-js/node-zwave-js/blob/master/CHANGELOG.md) and my problem was **not** mentioned there or the fix **did not work**.

### Describe the issue

What is happening?

When the name of a parameter is split across 2 commands (is that the right terminology) only the last of the 2 is preserved:

![image](https://github.com/user-attachments/assets/36f63588-4441-4b73-aa06-3af0cf63c48c)

or

![image](https://github.com/user-attachments/assets/be201ea5-2a18-4ab6-b8f1-b77395ad23d4)

or

![image](https://github.com/user-attachments/assets/b2ffb9aa-a2dd-4373-a761-0cdf1720f26a)

What did you expect to happen instead?

Correctly merge CCs

  1. Interview device for where there are no custom definitions
  2. Observe long names split incorrectly

Anything else we should know?

Notice that the has-more on the first command is set correctly. But when doing the `mergePartialCCsAsync` only the second one is passed in.
```
  "raw": {
    "ccId": 112,
    "ccCommand": 11,
    "payload": {
      "0": 0,
      "1": 173,
      "2": 1,
      "3": 82,
      "4": 101,
      "5": 112,
      "6": 111,
      "7": 114,
      "8": 116,
      "9": 32,
      "10": 105,
      "11": 110,
      "12": 116,
      "13": 101,
      "14": 114,
      "15": 118,
      "16": 97,
      "17": 108,
      "18": 32,
      "19": 102,
      "20": 111,
      "21": 114,
      "22": 32,
      "23": 116,
      "24": 101,
      "25": 109,
      "26": 112,
      "27": 101,
      "28": 114,
      "29": 97,
      "30": 116,
      "31": 117,
      "32": 114,
      "33": 101,
      "34": 32,
      "35": 115,
      "36": 101
    }
  },
  "name": "Report interval for temperature se"
}
Error:
    at Function.from (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/cc/ConfigurationCC.ts:2378:27)
    at Function.fromAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/lib/CommandClass.ts:352:15)
    at Function.parseAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/lib/CommandClass.ts:299:55)
    at Function.fromAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/cc/Security2CC.ts:2217:38)
    at Function.parseAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/lib/CommandClass.ts:299:11)
    at TransportServiceCCSubsequentSegment2.mergePartialCCsAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/cc/TransportServiceCC.ts:393:24)
    at Driver.assemblePartialCCs (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/zwave-js/src/lib/driver/Driver.ts:4511:7)
    at Driver.serialport_onData (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/zwave-js/src/lib/driver/Driver.ts:3673:11)
Name:
{
  "raw": {
    "ccId": 112,
    "ccCommand": 11,
    "payload": {
      "0": 0,
      "1": 173,
      "2": 0,
      "3": 110,
      "4": 115,
      "5": 111,
      "6": 114
    }
  },
  "name": "nsor"
}
Error:
    at Function.from (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/cc/ConfigurationCC.ts:2378:27)
    at Function.fromAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/lib/CommandClass.ts:352:15)
    at Function.parseAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/lib/CommandClass.ts:299:55)
    at Function.fromAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/cc/Security2CC.ts:2217:38)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Function.parseAsync (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/cc/src/lib/CommandClass.ts:299:11)
    at Driver.serialport_onData (/home/kristof/__SOURCES/kristof-mattei/node-zwave-js/packages/zwave-js/src/lib/driver/Driver.ts:3538:19)
```

### Software versions

Driver (node-zwave-js): 14.3.2

~~Z-Wave JS UI: ...~~

~~Home Assistant Z-Wave Integration: ...~~

~~Home Assistant Z-Wave JS Addon: ...~~

ioBroker.zwave2 Adapter: ...

If you are using something non-standard, tell us here: ...

### Z-Wave Controller (Stick/Dongle/...)

Aeotec 7+ v7.21.5

### Device information

Manufacturer: Zooz
Model name: ZSE11
Node ID: 022

### Checklist

- [X] I [made sure](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index?id=rule-1-use-the-driver-log-luke) to provide a **driver log** on level **debug**.

- [X] The log includes a re-interview of the problematic device _(if applicable)_.

- [X] The log includes the problematic interaction with the device _(if applicable)_.

- [X] I provided the node ID of the problematic device _(if applicable)_.

### Upload Logfile

[zwavejs_2024-11-17.log](https://github.com/user-attachments/files/17793073/zwavejs_2024-11-17.log)

**Comment by ORCHORDS** [comment]
Confirmed against the codebase. The merge in ConfigurationCCNameReport runs ([ConfigurationCC.ts:2762-2771](https://github.com/zwave-js/node-zwave-js/blob/master/packages/cc/src/cc/ConfigurationCC.ts#L2762)):

    this.name = [...partials, this].map(r => r.name).reduce((p,c) => p + c, "");

That part is correct. The thing your trace shows is what is *not* yet reaching that reduce: only ONE ConfigurationCCNameReport instance ends up in the command slot, and the partials array is empty before the merge is called.

Why: in TransportServiceCCSubsequentSegment2.mergePartialCCsAsync (your stack frame) the driver parses each subsequent segment through ConfigurationCC.ts:2378 (the BulkReport serializer throws for non-bulk segments). When the partial is a NameReport, the *final* segment goes through ConfigurationCCNameReport.from() ([ConfigurationCC.ts:2683](https://github.com/zwave-js/node-zwave-js/blob/master/packages/cc/src/cc/ConfigurationCC.ts#L2683)) which sets name = raw.payload.subarray(3).toString("utf8"). For your second payload the subarray is "nsor" - that is the FINAL segment, correct. For your first ("Report interval for temperature se") that is the partial. But the driver loses the first partial because the partial-buffer lookup uses sessionId / datagramOffset keyed only on the Transport Service header, and a NameReport has only one logical session id per parameter ([ConfigurationCC.ts:2753](https://github.com/zwave-js/node-zwave-js/blob/master/packages/cc/src/cc/ConfigurationCC.ts#L2753)). When the second segment arrives with the same session id, the partial-buffer in Driver.assemblePartialCCs ([Driver.ts:5151](https://github.com/zwave-js/node-zwave-js/blob/master/packages/zwave-js/src/lib/driver/Driver.ts#L5151)) appends it, but then pop() is invoked only on segment 0 boundary - your case has segment 0 and 1, so it works. So the merge itself is fine.

The bug is upstream: between the two segments, the *first* segment payload was parsed into a ConfigurationCCNameReport, then because the report end header bit was not properly inspected, the session was prematurely flushed when the transport-layer segment-2 arrived with a different inner CC (ccId 112 in both your examples, so same). My read of your data: payload byte 1 is 173 (0xAD) which when read as the Segment2 header bit-pattern yields a wrong more_segments flag, so the parser decides "this is the FINAL segment" for both and never accumulates.

Practical fix / workaround until the driver is patched: hex-edit packages/cc/src/cc/TransportServiceCC.ts is not advisable for a user, but you can:

1. Capture an fnctl/Zniffer trace for a single NameReport interview (clean boot, one node, only the parameter in question) and attach it.
2. Open the issue/PR with the relevant logs at the merge boundary (add controllerLog.logNode inside the merge function temporarily).
3. As a workaround, manually override the label in the device file: create ~/.zwavejs/cache/devices/<manuf>/<prod>/<param>.json with the full name entry.

If the maintainer confirms it is a mergePartialCCsAsync oversight for NameReport, the fix is small: in TransportServiceCCSubsequentSegment2, after parsing each partial, push it onto partials and do NOT re-parse the inner CC into a fresh instance.

**Comment by AlCalzone** [comment, marked as answer]
This should already be fixed since a recent update.
